sumcheck/src/util.rs

```diff
@@ -140,73 +136,175 @@ fn inner_extrapolate<F: Field, const IS_PARALLEL: bool>(
+/// w_j = 1 / ∏_{k ≠ j} (x_j - x_k)
+/// these weights are used in the barycentric form of Lagrange interpolation, which allows
+/// for efficient evaluation of the interpolating polynomial at any other point
+/// The weights depend only on the interpolation nodes and can be precomputed once
+fn extrapolate_uni_poly_deg_1<F: Field>(p_i: &[F; 2], eval_at: F) -> F {
+    let x0 = F::ZERO;
+    let x1 = F::ONE;
+
+    // w0 = 1 / (0−1) = -1
+    // w1 = 1 / (1−0) =  1
+    let w0 = -F::ONE;
+    let w1 = F::ONE;
+
+    let d0 = eval_at - x0;
+    let d1 = eval_at - x1;
+
+    let inv_d0 = d0.inverse();
+    let inv_d1 = d1.inverse();
+
+    let t0 = w0 * p_i[0] * inv_d0;
+    let t1 = w1 * p_i[1] * inv_d1;
+
+    let b0 = w0 * inv_d0;
+    let b1 = w1 * inv_d1;
+
+    (t0 + t1) * (b0 + b1).inverse()
 }
 
-/// compute the factorial(a) = 1 * 2 * ... * a
-#[inline]
-fn field_factorial<F: Field>(a: usize) -> F {
-    let mut res = F::ONE;
-    for i in 2..=a {
-        res *= F::from_u64(i as u64);
+fn extrapolate_uni_poly_deg_2<F: Field>(p_i: &[F; 3], eval_at: F) -> F {
+    let x0 = F::from_u64(0);
+    let x1 = F::from_u64(1);
+    let x2 = F::from_u64(2);
+
+    // w0 = 1 / ((0−1)(0−2)) =  1/2
+    // w1 = 1 / ((1−0)(1−2)) = -1
+    // w2 = 1 / ((2−0)(2−1)) =  1/2
+    let w0 = F::from_u64(1).div(F::from_u64(2));
+    let w1 = -F::ONE;
+    let w2 = F::from_u64(1).div(F::from_u64(2));
+
+    let d0 = eval_at - x0;
+    let d1 = eval_at - x1;
+    let d2 = eval_at - x2;
+
+    let inv_d0 = d0.inverse();
+    let inv_d1 = d1.inverse();
+    let inv_d2 = d2.inverse();
+
+    let t0 = w0 * p_i[0] * inv_d0;
+    let t1 = w1 * p_i[1] * inv_d1;
+    let t2 = w2 * p_i[2] * inv_d2;
+
+    let b0 = w0 * inv_d0;
+    let b1 = w1 * inv_d1;
+    let b2 = w2 * inv_d2;
+
+    (t0 + t1 + t2) * (b0 + b1 + b2).inverse()
+}
+
+fn extrapolate_uni_poly_deg_3<F: Field>(p_i: &[F; 4], eval_at: F) -> F {
+    let x0 = F::from_u64(0);
+    let x1 = F::from_u64(1);
+    let x2 = F::from_u64(2);
+    let x3 = F::from_u64(3);
+
+    // w0 = 1 / ((0−1)(0−2)(0−3)) = -1/6
+    // w1 = 1 / ((1−0)(1−2)(1−3)) =  1/2
+    // w2 = 1 / ((2−0)(2−1)(2−3)) = -1/2
+    // w3 = 1 / ((3−0)(3−1)(3−2)) =  1/6
+    let w0 = -F::from_u64(1).div(F::from_u64(6));
+    let w1 = F::from_u64(1).div(F::from_u64(2));
+    let w2 = -F::from_u64(1).div(F::from_u64(2));
+    let w3 = F::from_u64(1).div(F::from_u64(6));
+
+    let d0 = eval_at - x0;
+    let d1 = eval_at - x1;
+    let d2 = eval_at - x2;
+    let d3 = eval_at - x3;
+
+    let inv_d0 = d0.inverse();
+    let inv_d1 = d1.inverse();
+    let inv_d2 = d2.inverse();
+    let inv_d3 = d3.inverse();
+
+    let t0 = w0 * p_i[0] * inv_d0;
+    let t1 = w1 * p_i[1] * inv_d1;
+    let t2 = w2 * p_i[2] * inv_d2;
+    let t3 = w3 * p_i[3] * inv_d3;
+
+    let b0 = w0 * inv_d0;
+    let b1 = w1 * inv_d1;
+    let b2 = w2 * inv_d2;
+    let b3 = w3 * inv_d3;
+
+    (t0 + t1 + t2 + t3) * (b0 + b1 + b2 + b3).inverse()
```

Comment:
According to barycentric formula, `p(r) = \sum_i p[i] * w[i] * prod / eval_at - i)` where `prod = ∏ (eval_at - i)`;

Since `t[i] = w[i] * p[i] * 1 / (eval_at - i)`, then does it mean `1 / (b0 + b1 + b2 + b3) = prod`???

Reply:
If we take constant poly `f(x) = 1`, then `1 = f(eval_at) = \sum_i w[i] * 1 * prod / (eval_at - i)`. 
If we take `prod` out from the right hand in above equation, we get `prod = 1 / (\sum_i w[i] * b[i])`.

Reply:
@hero78119 Can you add the above explanation to the comments?